### PR TITLE
[ASAN] Intercept urProgramLink in sanitizer layer

### PR DESCRIPTION
**Pull request by jinge90**
Openmp offloading runtime (libomptarget) calls urProgramLink to build and link user's device code, this API needs to be intercepted to register device globals.
intel/llvm: https://github.com/intel/llvm/pull/13048

### CODE REVIEW

**Comment by jinge90** [comment]
Hi, @zhaomaosu 
could you help review this patch?
Thanks very much.

**Review by zhaomaosu — APPROVED** [review]
lgtm

**Comment by jinge90** [comment]
Hi, @oneapi-src/unified-runtime-maintain 
Could you help review this patch?
Thanks very much.

**Review by pbalcer — APPROVED** [review]
lgtm, but please separate out the unrelated changes.

File: source/loader/layers/sanitizer/asan_interceptor.cpp [hunk 1]
```diff
@@ -537,14 +537,28 @@ ur_result_t SanitizerInterceptor::enqueueAllocInfo(
     }
 
     // Left red zone
-    UR_CALL(enqueueMemSetShadow(Context, Device, Queue, AllocInfo->AllocBegin,
-                                AllocInfo->UserBegin - AllocInfo->AllocBegin,
-                                ShadowByte, LastEvent, &LastEvent));
+    uptr LeftRZSize = AllocInfo->UserBegin - AllocInfo->AllocBegin;
```
[the remark this answers is not in the document]
Comment (jinge90) [reply source/loader/layers/sanitizer/asan_interceptor.cpp:540]: Hi, @pbalcer 
After intercepting urProgramLinkExp, device sanitizer for omp offloading can run on GPU and a small bug exposed, we fixed in the same PR.
Thanks very much.

[the remark this answers is not in the document]
Comment (pbalcer) [reply source/loader/layers/sanitizer/asan_interceptor.cpp:540]: It's fine, but make it a separate commit.

[the remark this answers is not in the document]
Comment (jinge90) [reply source/loader/layers/sanitizer/asan_interceptor.cpp]: Hi, @pbalcer 
I just reverted the bug fix and will submit another patch, could you help review again?
Thanks very much.

**Comment by pbalcer** [comment]
It's very unlikely that this will affect anything on intel/llvm, right? Despite that, please follow the process outlined here: https://oneapi-src.github.io/unified-runtime/core/CONTRIB.html#adapter-change-process and create an intel/llvm PR.

File: source/loader/layers/sanitizer/asan_interceptor.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (pbalcer) [reply source/loader/layers/sanitizer/asan_interceptor.cpp]: Looks good, thanks!

**Comment by jinge90** [comment]
Hi, @pbalcer 
intel/llvm pre-ci has passed: https://github.com/intel/llvm/pull/13048 , could you help merge this ur patch?
Thanks very much.

**Review by AllanZyne — APPROVED** [review]
LGTM

**Comment by jinge90** [comment]
> > Hi, @pbalcer
> > intel/llvm pre-ci has passed: [intel/llvm#13048](https://github.com/intel/llvm/pull/13048) , could you help merge this ur patch?
> > Thanks very much.
> 
> We need to make sure about not conflicting with the other PR's in our merge queue.

Hi, @kbenzie 
May I know the status about this PR? Is there anything I need to do to merge the PR?
Thanks very much.

**Comment by kbenzie** [comment]
> Hi, @kbenzie
> May I know the status about this PR? Is there anything I need to do to merge the PR?
> Thanks very much.

I think its in a good state, we've still got a few [v0.9.x](https://github.com/oneapi-src/unified-runtime/pulls?q=is%3Aopen+is%3Apr+label%3Av0.9.x) PR's to merge but I hope to be done with those soon.

**Comment by jinge90** [comment]
Hi, @pbalcer and @kbenzie 
I just updated the patch to align with latest main branch, the related intel/llvm pr is updated as well, could you help review again?
Thanks very much!

**Pull request merged**